src/main/java/net/glowstone/util/compiler/EvalTask.java

```diff
@@ -0,0 +1,185 @@
+        ClassDataFileManager classDataFileManager = new ClassDataFileManager(compiler
+            .getStandardFileManager(null, null, null));
+
+        List<JavaFileObject> compilationUnit = new ArrayList<>();
+        compilationUnit.add(new JavaSource("REPLShell", source));
+
+        JavaCompiler.CompilationTask task = compiler.getTask(null, classDataFileManager,
+            null, null, null, compilationUnit);
+        if (task.call()) {
+            try {
+                Object returned = MethodInvocationUtils
+                    .invokeStaticMethod(getCompiledClass("REPLShell"), "run",
+                        lastOutput);
+                if (output) {
+                    lastOutput = returned;
+                    GlowServer.logger.info(" -> " + returned);
+                } else {
+                    GlowServer.logger.info(" -> <no value>");
+                }
+            } catch (Exception e) {
+                GlowServer.logger.log(Level.SEVERE, "Error in running REPL shell! ", e);
+            }
+        }
+    }
+
+    /**
+     * Find a class compiled in this tasks' class loader.
+     *
+     * @param name The name of the class to find.
+     * @return The class.
+     */
+    public Class<?> getCompiledClass(String name) {
+        return classLoader.findClass(name);
+    }
+
+    private class JavaSource extends SimpleJavaFileObject {
+
+        private final String source;
+
+        JavaSource(String name, String source) {
+            super(URI.create("string:///" + name + ".java"), Kind.SOURCE);
+            this.source = source;
+        }
+
+        @Override
+        public CharSequence getCharContent(boolean ignoreEncodingErrors) throws IOException {
+            return source;
+        }
+    }
+
+    private class MapClassLoader extends ClassLoader {
+
+        @Override
+        protected Class<?> findClass(String name) {
+            return defineClass(name, classData.get(name), 0, classData.get(name).length);
+        }
+    }
+
+    public class JavaClass extends SimpleJavaFileObject {
+
+        private final String name;
+
+        public JavaClass(String name) {
+            super(URI.create("string:///" + name + ".java"), Kind.CLASS);
+            this.name = name;
+        }
+
+        @Override
+        public OutputStream openOutputStream() throws IOException {
+            return new ClassDataOutputStream(name);
+        }
+    }
+
+    public class ClassDataFileManager extends ForwardingJavaFileManager<StandardJavaFileManager> {
+
+        public ClassDataFileManager(StandardJavaFileManager fileManager) {
+            super(fileManager);
+        }
+
+        @Override
+        public JavaFileObject getJavaFileForOutput(Location location, String className,
+            JavaFileObject.Kind kind, FileObject sibling) throws IOException {
+            return new JavaClass(className);
+        }
+    }
+
+    public class ClassDataOutputStream extends OutputStream {
```

Comment:
What does this do differently than the underlying ByteArrayOutputStream? If this can't simply be replaced by a BAOS, document why not, and can we use a FilterOutputStream instead?